### ISSUE TITLE
widget-x2many-2d-matrix 10.0.1.0.0.99.dev2

### ISSUE THREAD

**Issue by samersarhan** [OCA/web#674]
Some thing wrong.
the x_field and y_filed in my related model are `M2O,` in the 2d-matrix widget, it shows the id of M2O, not the name ex.product_id is x field but the showing is the id field value, not the name of the product template as it supposes to work.
in addition to this when I follow the read my file it says the following 

`
<field name="task_ids" widget="x2many_2d_matrix" field_x_axis="project_id" field_y_axis="user_id" field_value="planned_hours">`
  `   <tree>`
         `<field name="task_ids"/>`
         `<field name="project_id"/>`
        ` <field name="user_id"/>`
        ` <field name="planned_hours"/>`
     `</tree>`
 `</field> 
`
which also gives an error because the task_ids not a field in the co model

any help.
![capture](https://user-images.githubusercontent.com/16674144/28261145-b500cc0c-6ae6-11e7-8645-a8f6a22c3fa4.JPG)

**Comment by samersarhan** [comment]
@llacroix 
@rvalyi 
@zeroheure 
@nbessi

**Comment by samersarhan** [comment]
It solved by adding x_name and y_name as a Char field, but is this widget work only for wizards?
since we have to provide a dataset each time?

**Comment by norlinhenrik** [comment]
I can confirm that the documentation example
> gives an error because the task_ids not a field in the co model

When I remove task_ids from the tree I get this:
![image](https://user-images.githubusercontent.com/11876900/28943604-afbdb590-789f-11e7-9f47-df999599b5ba.png)

@samersarhan how did you create the matrix?

**Comment by samersarhan** [comment]
please clear your question. did you mean how to provide data to the matrix widget? 
if this is your question you should provide the data each time you run the widget by use default_get, on change or by linking to a function

**Comment by norlinhenrik** [comment]
Yes I mean how to provide data to the matrix widget? I tried the example that you posted from the documentation, but as you say id doesn't work...

**Comment by norlinhenrik** [comment]
All I want is an example that works. The example in the [documentation](https://github.com/OCA/web/tree/10.0/web_widget_x2many_2d_matrix) doesn't.

**Comment by samersarhan** [comment]
please, share your code, cause I followed the documentation so maybe you are missing something,

**Comment by norlinhenrik** [comment]
@samersarhan, in the first post you posted some xml that "gives an error because the task_ids not a field in the co model". How did you fix this error?

**Comment by norlinhenrik** [comment]
Here is the code for module task_matrix that gives an error:

manifest.py

```
# -*- coding: utf-8 -*-
{
    'name': "Task Matrix",
    'depends': ['project', 'web_widget_x2many_2d_matrix'],
    'data': ['views.xml'],
}
```

models.py

```
# -*- coding: utf-8 -*-
from odoo import fields, models

class MyWizard(models.TransientModel):
   _name = 'my.wizard'

   def _default_task_ids(self):
       # your list of project should come from the context, some selection
       # in a previous wizard or wherever else
       projects = self.env['project.project'].browse([1, 2, 3])
       # same with users
       users = self.env['res.users'].browse([1, 2, 3])
       return [
           (0, 0, {
               'project_id': p.id,
               'user_id': u.id,
               'planned_hours': 0,
               'message_needaction': False,
               'date_deadline': fields.Date.today(),
           })
           # if the project doesn't have a task for the user, create a new one
           if not p.task_ids.filtered(lambda x: x.user_id == u) else
           # otherwise, return the task
           (4, p.task_ids.filtered(lambda x: x.user_id == u)[0].id)
           for p in projects
           for u in users
       ]

   task_ids = fields.Many2many('project.task', default=_default_task_ids)
```


views.xml

```
<?xml version="1.0" encoding="utf-8"?>
<odoo>
    <data>
		<record model="ir.ui.view" id="view_task_matrix">
    		<field name="name">Task Matrix</field>
    		<field name="model">my.wizard</field>
    		<field name="priority" eval="20"/>
    		<field name="arch" type="xml">
				<form string="Task Matrix">
                    <field name="task_ids" widget="x2many_2d_matrix" field_x_axis="project_id" field_y_axis="user_id" field_value="planned_hours">
                        <tree>
                            <field name="task_ids"/>
                            <field name="project_id"/>
                            <field name="user_id"/>
                            <field name="planned_hours"/>
                        </tree>
                    </field>
				</form>
    		</field>
		</record>
		<record model="ir.actions.act_window" id="action_task_matrix">
    		<field name="name">Task Matrix</field>
    		<field name="res_model">my.wizard</field>
    		<field name="view_mode">form</field>
            <field name="view_id" ref="view_task_matrix"/>
		</record>
		<menuitem id="menu_task_matrix" name="Task Matrix" sequence="20" parent="project.menu_project_report" action="action_task_matrix"/>
    </data>
</odoo>
```

Error during installation (with project 10.0.1.1 and web_widget_x2many_2d_matrix 10.0.1.0.0)

```
ParseError: "Error while validating constraint

Field `task_ids` does not exist

Error context:
View `Task Matrix`
[view_id: 378, xml_id: n/a, model: my.wizard, parent_id: n/a]
None" while parsing file:///C:/Users/Henrik/Documents/odoo_modules/task_matrix/views.xml:4, near
<record model="ir.ui.view" id="view_task_matrix">
    		<field name="name">Task Matrix</field>
    		<field name="model">my.wizard</field>
    		<field name="priority" eval="20"/>
    		<field name="arch" type="xml">
				<form string="Task Matrix">
                    <field name="task_ids" widget="x2many_2d_matrix" field_x_axis="project_id" field_y_axis="user_id" field_value="planned_hours">
                        <tree>
                            <field name="task_ids"/>
                            <field name="project_id"/>
                            <field name="user_id"/>
                            <field name="planned_hours"/>
                        </tree>
                    </field>
				</form>
    		</field>
		</record>
```

**Comment by samersarhan** [comment]
Just remove the task_ids from the `<tree>` in XML file, your tree must be
`<tree>
    <field name="project_id"/>
    <field name="user_id"/>
     <field name="planned_hours"/>
</tree>`
and no need to set  web_widget_x2many_2d_matrix in the depend list since it's only widget,

**Comment by norlinhenrik** [comment]
Thanks! I have now
- removed the xml task_ids
- set the python project variable to .browse([1])
- installed my module
- activated the example project

When I click the menu to se the task matrix I get a Odoo Client Error:
```
Uncaught TypeError: Cannot read property 'id' of undefined
http://localhost:8100/web/content/262-31d1ffd/web.assets_backend.js:3870
Traceback:
TypeError: Cannot read property 'id' of undefined
    at Class.get_xy_id (http://localhost:8100/web/content/262-31d1ffd/web.assets_backend.js:3870:29)
    at eval (eval at _render (http://localhost:8100/web/content/251-0439e10/web.assets_common.js:2949:73), <anonymous>:50:74)
    at Object.foreach (http://localhost:8100/web/content/251-0439e10/web.assets_common.js:2919:18)
    at eval (eval at _render (http://localhost:8100/web/content/251-0439e10/web.assets_common.js:2949:73), <anonymous>:40:24)
    at Object.foreach (http://localhost:8100/web/content/251-0439e10/web.assets_common.js:2919:18)
    at Engine.eval (eval at _render (http://localhost:8100/web/content/251-0439e10/web.assets_common.js:2949:73), <anonymous>:32:23)
    at Engine._render (http://localhost:8100/web/content/251-0439e10/web.assets_common.js:2948:296)
    at Engine.render (http://localhost:8100/web/content/251-0439e10/web.assets_common.js:2948:151)
    at Engine._render (http://localhost:8100/web/content/251-0439e10/web.assets_common.js:2952:57)
    at Engine.render (http://localhost:8100/web/content/251-0439e10/web.assets_common.js:2948:151)
```

Any solution?

**Comment by samersarhan** [comment]
Please make sure you have projects with ids[1,2,3] and users with ids[1,2,3],
or for the test in .py, you can set the following
```
users = self.env['res.users'].search([])
projects = self.env[''project.project'].search([])
```

**Comment by norlinhenrik** [comment]
The .search([]) methods fixed the error message. Thank you @samersarhan !

Now my matrix looks like this:
![image](https://user-images.githubusercontent.com/11876900/29260042-742baa7a-80c7-11e7-8732-54996683a138.png)

I created a new project, and then I got the above error again (Uncaught TypeError: Cannot read property 'id' of undefined). Then I created a task in the new project, and then the matrix looked just like on the image, with no changes.

What is the next step to show the matrix correctly? Here is my module:
[task_matrix.zip](https://github.com/OCA/web/files/1221389/task_matrix.zip)
